### PR TITLE
MergeTree read apply LIMIT for limit only queries

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
MergeTree read apply LIMIT for limit only queries.

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 2061779f0adc23bf2c550cab0849a66d0711161f with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58504/2061779f0adc23bf2c550cab0849a66d0711161f/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58504/2061779f0adc23bf2c550cab0849a66d0711161f/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 1]
```diff
@@ -1161,6 +1161,77 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
...
+
+        auto ranges_it = result_part_with_ranges.ranges.begin();
+        auto ranges_end_it = result_part_with_ranges.ranges.end();
+        for (; ranges_it != ranges_end_it; ++ranges_it)
+        {
+            const auto & range = *ranges_it;
+            if (remaining_limit == 0)
+                break;
+
+            size_t rows_count_in_range = part_index_granularity.getRowsCountInRange(range);
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:1196]: How does it work with lightweight deletes? I suspect that this method can return the number of rows together with deleted. Let's check it.

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 2]
```diff
@@ -1161,6 +1161,77 @@ RangesInDataParts MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipInd
...
+            while (left < right)
+            {
+                size_t middle = left + (right - left) / 2;
+                size_t middle_value = marks_rows_partial_sums[middle] - mark_range_partial_sum_start;
+
+                if (remaining_limit > middle_value)
+                    left = middle + 1;
+                else
+                    right = middle;
+            }
```
Comment (KochetovNicolai) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:1217]: bin search :) maybe replace it with std::lower_bound or something?

**Review by KochetovNicolai — COMMENTED** [review]
Idea looks ok

**Comment by kitaisreal** [comment]
Need to check performance tests slowdown https://s3.amazonaws.com/clickhouse-test-reports/58504/b26b791720f01d8311108a986b0c0578c3b38024/performance_comparison_[1_4]/report.html.

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:877]: Updated.

File: src/Processors/QueryPlan/ReadFromMergeTree.cpp [hunk 3]
```diff
@@ -437,19 +438,19 @@ Pipe ReadFromMergeTree::readFromPool(
...
-            storage_snapshot,
-            prewhere_info,
-            actions_settings,
-            reader_settings,
-            required_columns,
-            virt_column_names,
-            pool_settings,
-            context);
-    }
+    // else
```
Comment (alexey-milovidov) [inline src/Processors/QueryPlan/ReadFromMergeTree.cpp:441]: @kitaisreal, it's not allowed to have code commented out.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

**Comment by alexey-milovidov** [comment]
@KochetovNicolai, if the idea is ok, could you please open another PR and continue?

**Pull request closed without merging**